master/internal/resourcemanagers/fitting_methods.go

```diff
@@ -12,20 +12,62 @@ func labelSatisfied(req *AllocateRequest, agent *agentState) bool {
 	return req.Label == agent.label
 }
 
+func maxZeroSlotContainersSatisfied(req *AllocateRequest, agent *agentState) bool {
+	if req.SlotsNeeded == 0 {
+		if agent.maxZeroSlotContainers == nil {
+			return true
+		}
+		if *agent.maxZeroSlotContainers == 0 {
+			return false
+		}
+		return agent.numZeroSlotContainers() < *agent.maxZeroSlotContainers
+	}
+	return true
+}
+
+func agentIdleSatisfied(_ *AllocateRequest, agent *agentState) bool {
+	return agent.idle()
+}
+
 // Soft Constraints
 
 // BestFit returns a float affinity score between 0 and 1 for the affinity between the task and
 // the agent. This method attempts to allocate tasks to the agent that is both most utilized and
 // offers the fewest slots. This method should be used when the cluster is dominated by multi-slot
 // applications.
-func BestFit(_ *AllocateRequest, agent *agentState) float64 {
+func BestFit(req *AllocateRequest, agent *agentState) float64 {
+	if req.SlotsNeeded == 0 {
+		if agent.maxZeroSlotContainers == nil {
+			if agent.numZeroSlotContainers() == 0 {
+				return 0
```

Comment:
This is zero because the agent isn't running any zero slot containers and we don't want to schedule any tasks onto that container? I think we should consider this scenario:
- Two agents with 4 GPUs each and `maxZeroSlotContainers = nil`
- AgentA is running a 4 GPU container
- AgentB is running a zero-slot Tensorboard
- We try to schedule a new Tensorboard

With best fit, I believe the desired end result is that we have the GPU task and both Tensorboards running on the same agent. Won't this implementation prevent that?

Reply:
> This is zero because the agent isn't running any zero slot containers and we don't want to schedule any tasks onto that container?

Bestfit wants the fitting score of an occupied agent to be higher than the score of an idle agent. In the case of no limit for zero slot containers and zero-slot task, we simply return either 0 for idle agents or 1 for occupied agents. 0 doesn't mean the task can't be scheduled. It just fails to beat all the agents with a score 0.

> With best fit, I believe the desired end result is that we have the GPU task and both Tensorboards running on the same agent. Won't this implementation prevent that?

You are right. This implementation doesn't prevent that. That's why people should set `maxCPUContainersPerAgent` of their gpu pools to be 0.

Reply:
My question is more - why is the score 0 instead of taking into account the number of non-CPU containers running?

>That's why people should set maxCPUContainersPerAgent of their gpu pools to be 0.

I don't think we can expect them to do this.